Developer/bjs2 77887

SAM19991 opened the pull request:
Анализ комментариев
Задание
analytics_service должен собирать информацию о комментариях поста пользователя/проекта другими юзерами в post_service:

Создать Spring-конфигурацию Redis топика и реализацию CommentEventPublisher — отправителя события CommentEvent в post_service. Туториал: [PubSub Messaging with Spring Data Redis | Baeldung](https://www.baeldung.com/spring-data-redis-pub-sub)  .

Отправлять событие CommentEvent через CommentEventPublisher в момент, когда пользователь комментит пост другого пользователя — вызывает соответствующий эндпоинт. В событии содержатся: id поста; id автора; id коммента; дата и время.

В analytics_service создать конфигурацию слушателя эвентов CommentEventListener для CommentEvent, руководствуясь тем же тутором выше.

CommentEventListener слушает события CommentEvent и сохраняет информацию из этих событий с БД, используя AnalyticsEventService для выполнения логики сохранения и AnalyticsEvent для превращения CommentEvent в сущность БД. Используем AnalyticsEventMapper для такого превращения.

Критерии приема
Сообщение о новом просмотре профиля отправляется в новый топик Redis из post_service.

Сообщение потребляется из нового топика в Redis в analytics_service.

Данные о событии сохраняются в БД в соответствующую таблицу analytics_event.

Всё покрыто unit-тестами.

kashkademan committed:
Solution(1). 77891 Save and get analytics.

kashkademan committed:
Fix(1). 77891 Save and get analytics. Team's review.

SAM19991 committed:
Merge remote-tracking branch 'origin/triton-master-stream10' into triton-master-stream10

SAM19991 committed:
BJS2-77887

SAM19991 committed:
- Изменен путь конфигурации Redis-канала для комментариев
  с app.redis.topic.comment-analytics на spring.data.redis.channel.comment-analytics
- Добавлен соответствующий параметр в application.yaml
- Обеспечена единообразная структура конфигурации Redis-каналов

mysosnovskaya commented on src/main/java/faang/school/analytics/controller/CommentAnalyticsController.java:
```diff
@@ -0,0 +1,49 @@
+        Optional<Interval> intervalOpt = Optional.ofNullable(interval)
+                .map(intervalStr -> {
+                    try {
+                        return Interval.valueOf(intervalStr.toUpperCase());
```
Можно сразу Interval вместо String принимать, спринг так умеет)

mysosnovskaya commented on src/main/java/faang/school/analytics/mapper/AnalyticsEventMapper.java:
```diff
@@ -2,21 +2,33 @@
+
+    @Mapping(source = "postId", target = "receiverId")
+    @Mapping(source = "autorId", target = "actorId")
+    @Mapping(target = "eventType", expression = "java(EventType.POST_COMMENT.name())")
```
Тут можно использовать constant = "POST_COMMENT" вместо expression

mysosnovskaya commented on src/main/java/faang/school/analytics/listener/CommentEventListener.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public void onMessage(Message message, byte[] pattern) {
+        CommentEvent commentEvent;
```
Лучше внутрь try занести. Пока не вижу причин оставлять это снаружи

mysosnovskaya commented on src/main/java/faang/school/analytics/errorMessage/ErrorMessage.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Похоже на утилитный класс, подошла бы аннотация UtilityClass от lombok

SAM19991 committed:
Исправлены опечатки: `autorId` → , `comentEvent` → `commentEvent` `authorId`
 Упрощен маппинг eventType (`expression` → `constant`)
 Убрана избыточная логика в контроллере
 Рефакторинг: `@UtilityClass` вместо `@NoArgsConstructor`

MiraGlaceon commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -34,10 +38,13 @@ public RedisTemplate<String, Object> redisTemplate(RedisConnectionFactory connec
             RedisConnectionFactory connectionFactory,
             MessageListenerAdapter listenerAnalyticsEventAdapter,
-            ChannelTopic analyticsTopic) {
+            ChannelTopic analyticsTopic,
```
С ребятами решили проблему с расширением листенеров. Подключайся к нам на созвон в понедельник, будем обсуждать этот момент

MiraGlaceon commented on src/main/java/faang/school/analytics/controller/CommentAnalyticsController.java:
```diff
@@ -0,0 +1,42 @@
+            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime from,
+            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime to) {
+
+        Optional<Interval> intervalOpt = Optional.ofNullable(interval);
```
Валидацию лучше вынести в отдельный сервис-компонент и интегрировать с analyticsService или сразу в аналитикс сервисе провалидировать. Контроллер не должен содержать бизнес логики, только обрабатывать запросы клиента

MiraGlaceon commented on src/main/java/faang/school/analytics/errorMessage/ErrorMessage.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class ErrorMessage {
```
Тут лучше дать имя подробнее для класса. Потому что у тебя текст ошибки для комментариев, а имя класса абстрактно. Так же утилитные классы не должны содержать в себе бизнес-логики, это классы, которые выполняют тривиальные и независимые от бизнес-домена вещи. Например, конвертировать дату в строку или перевести поток байтов в другую кодировку и тп.
Предлагаю сразу сделать класс ошибки Exceptional с нужными константами и бросать эту ошибку сразу где нужно

MiraGlaceon commented on src/main/java/faang/school/analytics/listener/CommentEventListener.java:
```diff
@@ -0,0 +1,45 @@
+            log.info("Saved analytics event from comment: {}", analyticsEventDto);
+
+        } catch (IOException e) {
+            String errorMessage = ErrorMessage.formatDeserializationError(e.getClass().getSimpleName() + ": " + e.getMessage());
```
Вот здесь, к примеру, можно сразу в лог записать текст ошибки и не делать утилитный класс. Либо сразу создать ошибку и бросить ее. Попробуй избавиться от ErrorMessage и, либо преобразовать его в эксепшен, либо уложить все в класс CommentEventDeserializationException

SAM19991 committed:
Создан новый сервис `AnalyticsValidationService`
Валидация вынесена из контроллера в сервисный слой
Обновлен для использования валидации `AnalyticsService`

Убрана бизнес-логика из `CommentAnalyticsController`

Удален утилитный класс `ErrorMessage`
Обновлен `CommentEventDeserializationException`
Упрощен `CommentEventListener`
